sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java

```diff
@@ -439,4 +652,553 @@ public void process(ProcessContext c) {
+  public abstract static class Write<DestinationT, UserT>
+      extends PTransform<PCollection<UserT>, WriteFilesResult<DestinationT>> {
+    /**
+     * Information available for constructing a filename in {@link FilenamePolicy}. A {@link
+     * FilenamePolicy} MUST return unique filenames for different combinations of {@link
+     * #getNumShards}, {@link #getShardIndex}, {@link #getWindow}, {@link #getPane}.
+     */
+    interface FilenameContext {
+      /** Current window being written. */
+      BoundedWindow getWindow();
+
+      /** Pane of the current window being written. */
+      PaneInfo getPane();
+
+      /** Total number of shards generated for the current window and pane. */
+      int getNumShards();
+
+      /**
+       * Index of the current shard among all the shards generated for the current window and pane.
+       */
+      int getShardIndex();
+
+      /** The specified {@link Compression}. */
+      Compression getCompression();
+    }
+
+    /**
+     * A policy for generating names for shard files using information in {@link FilenameContext}.
+     */
+    interface FilenamePolicy extends Serializable {
+      /** Generates the filename. */
+      ResourceId getFilename(FilenameContext context);
+    }
+
+    /**
+     * Generates filenames by concatenating prefix and filling information in the shard template.
+     *
+     * <ul>
+     *   <li>N stands for {@link FilenameContext#getNumShards()}. NNN stands for formatting it into
+     *       3 digits, e.g. 042.
+     *   <li>S stands for {@link FilenameContext#getShardIndex()}. SSS stands for formatting it into
+     *       3 digits, e.g. 042.
+     *   <li>W stands for {@link FilenameContext#getWindow()}
+     *   <li>P stands for {@link FilenameContext#getPane()}
+     * </ul>
+     */
+    public static FilenamePolicy prefixAndShardTemplate(
+        final String prefix, final String shardTempate) {
+      final DefaultFilenamePolicy policy =
+          new DefaultFilenamePolicy(
+              new DefaultFilenamePolicy.Params()
+                  .withBaseFilename(FileSystems.matchNewResource(prefix, false /* isDirectory */))
+                  .withShardTemplate(shardTempate));
+      return new FilenamePolicy() {
+        @Override
+        public ResourceId getFilename(final FilenameContext context) {
+          return policy.windowedFilename(
+              context.getShardIndex(),
+              context.getNumShards(),
+              context.getWindow(),
+              context.getPane(),
+              FileBasedSink.CompressionType.fromCanonical(context.getCompression()));
+        }
+      };
+    }
+
+    /**
+     * Generates filenames as {@link #prefixAndShardTemplate(String, String)} with a shard template
+     * of {@link DefaultFilenamePolicy#DEFAULT_WINDOWED_SHARD_TEMPLATE}, which is {@code
+     * W-P-SSSSS-of-NNNNN}.
+     */
+    public static FilenamePolicy prefixAndWindowedShard(final String prefix) {
```

Comment:
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
the names of these methods are confusing - they don't tell me what they do.